lib/PrivateAppUserTokenManager.ts

```diff
@@ -0,0 +1,249 @@
+import {
+  fetchPrivateAppUserToken,
+  createPrivateAppUserToken,
+  updatePrivateAppUserToken,
+  PrivateAppUserTokenResponse,
+} from '../api/privateAppUserToken';
+import { scopeGroupsForPersonalAccessKey } from './personalAccessKey';
+import { logger } from './logger';
+import { i18n } from '../utils/lang';
+import moment from 'moment';
+import { AxiosError } from 'axios';
+import { getAxiosErrorWithContext } from '../errors/apiErrors';
+
+const USER_TOKEN_READ = 'developer.private_app.temporary_token.read';
+const USER_TOKEN_WRITE = 'developer.private_app.temporary_token.write';
+const i18nKey = 'lib.PrivateAppUserTokenManager';
+
+export class PrivateAppUserTokenManager {
+  accountId: number;
+  private tokenMap: Map<number, PrivateAppUserTokenResponse>;
+  private tokenMapIntervalId: Map<number, NodeJS.Timeout>;
+  private enabled: boolean;
+
+  constructor(accountId: number) {
+    this.accountId = accountId;
+    this.tokenMap = new Map<number, PrivateAppUserTokenResponse>();
+    this.tokenMapIntervalId = new Map<number, NodeJS.Timeout>();
+    this.enabled = false;
+  }
+
+  async init(): Promise<void> {
+    const scopeGroups = new Set<string>(
+      await scopeGroupsForPersonalAccessKey(this.accountId)
+    );
+    if (
+      !scopeGroups.has(USER_TOKEN_READ) ||
+      !scopeGroups.has(USER_TOKEN_WRITE)
+    ) {
+      logger.warn(
+        i18n(`${i18nKey}.errors.noScopes`, {
+          accountId: this.accountId,
+        })
+      );
+    } else {
+      logger.debug(
+        i18n(`${i18nKey}.enabled`, {
+          accountId: this.accountId,
+        })
+      );
+      this.enabled = true;
+    }
+  }
+
+  cleanup() {
+    this.tokenMapIntervalId.forEach(timeoutId => clearInterval(timeoutId));
+    this.tokenMapIntervalId.clear();
+    this.tokenMap.clear();
+  }
+
+  async getPrivateAppToken(
+    appId: number,
+    scopeGroups: string[] = []
+  ): Promise<string | undefined> {
+    if (!this.enabled) {
+      logger.debug(
+        i18n(`${i18nKey}.disabled`, {
+          accountId: this.accountId,
+          appId: appId,
+        })
+      );
+      return;
+    }
+    try {
+      if (
+        this.tokenMap.has(appId) &&
+        PrivateAppUserTokenManager.doesTokenHaveAllScopes(
```

Comment:
Could this check be switched to:
```js
if (this.doesTokenHaveAllScopes(appId, scopeGroups) { ... }
```

Reply:
update, can't just pass the appId since this method is used elsewhere when fetching the token from the api